### ISSUE TITLE
OTEL spans are not recognised as spans from a database

### ISSUE THREAD

**Issue by maxkomarychev** [prisma/prisma#21472]
### Bug description

When viewing spans in DataDog spans of Prisma are not recognised as database-ones.

![dd-screenshot](https://github.com/prisma/prisma/assets/7002833/df32f637-2d98-47bc-bd64-2e67c55f226b)

It appears like Prisma libraries do not match these conventions: https://opentelemetry.io/docs/specs/semconv/database/database-spans/ .

Specifically:

1. Span type is not set to "client"
2. attribute `db.system` is not populated

### How to reproduce

Enable open telemetry feature in Prisma and view spans in DataDog

### Expected behavior

Spans of Prisma should appear in "DB" section.

### Prisma information

```prisma
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider      = "prisma-client-js"
  // https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#binarytargets-options
  binaryTargets = ["darwin-arm64", "darwin", "debian-openssl-1.1.x", "linux-arm64-openssl-1.1.x"]
  previewFeatures = ["tracing"]
}

.... models go here ....

```

```ts
import { PrismaClient } from '@prisma/client'
import { PrismaInstrumentation } from '@prisma/instrumentation'
import { registerInstrumentations } from '@opentelemetry/instrumentation'

const prismaClient = new PrismaClient()

const instr = new PrismaInstrumentation({ middleware: true })
registerInstrumentations({
    instrumentations: [instr],
})

export { prismaClient }

```

### Environment & setup

- OS: docker - debian-bullseye
- Database: PostgreSQL
- Node.js version: 18

### Prisma Version

```
prisma                  : 4.11.0
@prisma/client          : 4.11.0
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.11.0-57.8fde8fef4033376662cad983758335009d522acb
Default Engines Hash    : 8fde8fef4033376662cad983758335009d522acb
Studio                  : 0.483.0
```

**Comment by aqrln** [comment]
I think it's fine to treat it as a bug if tools like DataDog rely on this information.

**Comment by AbhiPrasad** [comment]
Hey @aqrln is there a timeline around this? Happy to also PR this in, just need some guidance about how to discover `db.system` properly (allow the tracer to discover what the provider is)

**Comment by aqrln** [comment]
@AbhiPrasad I can't share any specific timeline yet but we do want to invest effort into tracing soon, so this issue will also be on the radar.

> just need some guidance about how to discover db.system properly (allow the tracer to discover what the provider is)

So, this actually depends a lot on one detail which is not clear to me. Do all Prisma spans need to match these conventions or only the one that represents the call to the database? From my understanding of the spec it seems to me that the latter is correct, but this issue, I believe, implies the former.

**Comment by AbhiPrasad** [comment]
The spans that represent calls to a database must have `db.system` on them, so not all the spans should be changed. I do think another look should be taken at for serialization/parsing related spans though.

**Comment by aqrln** [comment]
> The spans that represent calls to a database must have db.system on them, so not all the spans should be changed.

Awesome, that makes sense.

> I do think another look should be taken at for serialization/parsing related spans though.

Please let us know if you notice anything missing there!

> Happy to also PR this in, just need some guidance about how to discover db.system properly (allow the tracer to discover what the provider is)

If you are still willing to give it a try, an easy way would be to just pass specify it explicitly in every place the relevant spans are created since each connector knows what kind of database it is for.

So, for SQL databases it needs to be added here: https://github.com/prisma/prisma-engines/blob/793f7fb7404810e45c1f22d4b13a50cd4c8e0b9b/quaint/src/connector/metrics.rs#L12 and passed as a function argument from every caller of `metrics::query`.

For MongoDB it's scattered in a few places in these two files (grep for `"prisma:engine:db_query"`):
- https://github.com/prisma/prisma-engines/blob/793f7fb7404810e45c1f22d4b13a50cd4c8e0b9b/query-engine/connectors/mongodb-query-connector/src/root_queries/read.rs
- https://github.com/prisma/prisma-engines/blob/793f7fb7404810e45c1f22d4b13a50cd4c8e0b9b/query-engine/connectors/mongodb-query-connector/src/root_queries/write.rs

As for `postgres` connector being used for both PostgreSQL and CockroachDB, this is not a problem according to the spec:

> The actual DBMS may differ from the one identified by the client. For example, when using PostgreSQL client libraries to connect to a CockroachDB, the `db.system` is set to `postgresql` based on the instrumentation’s best knowledge.

however, if desired, can still be easily done: you can add a `flavour: PostgresFlavour` field in https://github.com/prisma/prisma-engines/blob/793f7fb7404810e45c1f22d4b13a50cd4c8e0b9b/quaint/src/connector/postgres/native/mod.rs#L48 and then save it from `url.flavour()` in `PostgreSql::new`.

**Comment by AbhiPrasad** [comment]
Got busy with some other tasks, but we're tracking this on the sentry side here: https://github.com/getsentry/sentry-javascript/issues/12234

Someone will take a look soon (and leave an update here accordingly)

**Comment by aqrln** [comment]
Related:
* https://github.com/prisma/prisma/issues/14640#issuecomment-1220097445
* https://github.com/prisma/prisma/issues/14640#issuecomment-1477786803
* https://github.com/prisma/prisma/issues/14640#issuecomment-1768299969

**Comment by jwd-dev** [comment]
Any updates here?

**Comment by Sarabadu** [comment]
just adding a related issue with AzureExporter also expecting this attribute
https://github.com/Azure/azure-sdk-for-js/issues/28065

**Comment by aqrln** [comment]
Hey folks, this should be fixed in the next release (5.22.0). Feel free to also try the latest `prisma@dev` and `@prisma/client@dev` if you want to test it before the stable release, and please let us know if you have any feedback or if we missed anything!
